superset/tasks/async_queries.py

```diff
@@ -0,0 +1,111 @@
+# under the License.
+
+import logging
+from typing import Any, cast, Dict, Optional
+
+from flask import current_app
+
+from superset import app
+from superset.exceptions import SupersetVizException
+from superset.extensions import async_query_manager, cache_manager, celery_app
+from superset.utils.cache import generate_cache_key, set_and_log_cache
+from superset.views.utils import get_datasource_info, get_viz
+
+logger = logging.getLogger(__name__)
+query_timeout = current_app.config[
+    "SQLLAB_ASYNC_TIME_LIMIT_SEC"
+]  # TODO: new config key
+
+
+@celery_app.task(name="load_chart_data_into_cache", soft_time_limit=query_timeout)
+def load_chart_data_into_cache(
+    job_metadata: Dict[str, Any], form_data: Dict[str, Any],
+) -> None:
+    from superset.charts.commands.data import (
+        ChartDataCommand,
+    )  # load here due to circular imports
+
+    with app.app_context():  # type: ignore
+        try:
+            command = ChartDataCommand()
+            command.set_query_context(form_data)
+            result = command.run(cache=True)
+            cache_key = result["cache_key"]
+            result_url = f"/api/v1/chart/data/{cache_key}"
+            async_query_manager.update_job(
+                job_metadata, async_query_manager.STATUS_DONE, result_url=result_url,
+            )
+        except Exception as exc:
+            # TODO: QueryContext should support SIP-40 style errors
+            error = exc.message if hasattr(exc, "message") else str(exc)  # type: ignore # pylint: disable=no-member
+            errors = [{"message": error}]
+            async_query_manager.update_job(
+                job_metadata, async_query_manager.STATUS_ERROR, errors=errors
+            )
+            raise exc
+
+        return None
+
+
+@celery_app.task(name="load_explore_json_into_cache", soft_time_limit=query_timeout)
+def load_explore_json_into_cache(
+    job_metadata: Dict[str, Any],
+    form_data: Dict[str, Any],
+    response_type: Optional[str] = None,
+    force: bool = False,
+) -> None:
+    with app.app_context():  # type: ignore
+        try:
+            datasource_id, datasource_type = get_datasource_info(None, None, form_data)
+
+            viz_obj = get_viz(
+                datasource_type=cast(str, datasource_type),
+                datasource_id=datasource_id,
+                form_data=form_data,
+                force=force,
+            )
+            # run query & cache results
+            payload = viz_obj.get_payload()
+            if viz_obj.has_error(payload):
+                raise SupersetVizException(errors=payload["errors"])
+
+            # cache form_data for async retrieval
+            cache_value = {"form_data": form_data, "response_type": response_type}
+            cache_key = generate_cache_key(
+                cache_value, "ejr-"
+            )  # ejr: explore_json request
+            set_and_log_cache(cache_manager.cache, cache_key, cache_value)
```

Comment:
Ditto `cache_manager.data_cache`.

Reply:
Data being cached here is from the client, and includes no query results. Happy to move this to `data_cache` if folks feel this is sensitive enough to warrant that.

Reply:
If it's just query context info then it probably doesn't have to.